cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,538 @@
+      default:
+        throw new IllegalArgumentException("Unsupported balancing action " + action.balancingAction() + " is provided.");
+    }
+  }
+
+  private boolean isSwapActionSelfSatisfied(ClusterModel clusterModel, BalancingAction action) {
+    if (action.balancingAction() != ActionType.INTER_BROKER_REPLICA_SWAP) {
+      throw new IllegalArgumentException("Expected INTER_BROKER_REPLICA_SWAP. But got: " + action.balancingAction());
+    }
+    Replica sourceReplica = clusterModel.broker(action.sourceBrokerId()).replica(action.topicPartition());
+    Replica destinationReplica = clusterModel.broker(action.destinationBrokerId()).replica(action.topicPartition());
+    if (sourceReplica.isLeader() == destinationReplica.isLeader() && Objects.equals(action.topic(), action.destinationTopic())) {
+      // Swapping either 2 leader replicas or 2 follower replicas of the same topic has no effect on this goal
+      return true;
+    }
+    return isSelfSatisfiedToMoveOneLeaderReplica(clusterModel.broker(action.sourceBrokerId()),
+                                                 clusterModel.broker(action.destinationBrokerId()),
+                                                 sourceReplica)
+           && isSelfSatisfiedToMoveOneLeaderReplica(clusterModel.broker(action.destinationBrokerId()),
+                                                    clusterModel.broker(action.sourceBrokerId()),
+                                                    destinationReplica);
+  }
+
+  private boolean actionAffectsRelevantTopics(BalancingAction action) {
+    if (_mustHaveTopicLeadersPerBroker.contains(action.topic())) {
+      return true;
+    }
+    return action.balancingAction() == ActionType.INTER_BROKER_REPLICA_SWAP &&
+           _mustHaveTopicLeadersPerBroker.contains(action.destinationTopic());
+  }
+
+  /**
+   * Check whether it is self-satisfied to move a leader replica of a topic with the given name from a the given source
+   * broker to the given destination broker
+   * @param sourceBroker the broker from which a leader replica is removed
+   * @param destinationBroker the broker from which a leader replica is added
+   * @param replicaToBeMoved replica to be moved from the source broker to the destination broker
+   * @return {@code true} if the given replica move would violate the optimized goal (i.e. the move is not acceptable),
+   * {@code false} otherwise.
+   */
+  private boolean isSelfSatisfiedToMoveOneLeaderReplica(Broker sourceBroker, Broker destinationBroker, Replica replicaToBeMoved) {
+    if (!replicaToBeMoved.isLeader()) {
+      return true;
+    }
+    String topicName = replicaToBeMoved.topicPartition().topic();
+    if (!_mustHaveTopicLeadersPerBroker.contains(topicName)) {
+      return true;
+    }
+    // Moving a replica leadership from a not-alive broker to an alive broker is self-satisfied
+    if (!sourceBroker.isAlive() && destinationBroker.isAlive()) {
+      return true;
+    }
```

Comment:
By design, this function should not be called with 
1. a non-leader replica -- please drop check on `if (!replicaToBeMoved.isLeader())`.
2. with replicas from topics other than `_mustHaveTopicLeadersPerBroker` -- please drop check on `if (!_mustHaveTopicLeadersPerBroker.contains(topicName))`.
3. offline replicas because this is already done in the proposed simplified version of `selfSatisfied`(also destinationBroker should not be dead if it is proposed here) -- please drop check on` if (!sourceBroker.isAlive() && destinationBroker.isAlive())`


-- See my comment above on the proposed simplified version of `selfSatisfied`.

Reply:
Good point!